### PR TITLE
Add code to setup environment and create base file

### PR DESCRIPTION
**Pull request by alice-hannah**
Hi all,

I've added you all as reviewers as not sure who has time/plans to do this - no pressure for you all to review it!

Some notes:
* As well as the variables required for the publication, we are also including the following in the base file; quarter, locality, ca2018 and hscp2018. These are so that the code/file can also be used to produce  data for National Indicator reporting in Source.
* Open to ideas on a better way to exclude external causes of death
* I'm aware the extracted data isn't being saved anywhere, however I'm imagining that the data preparation code will eventually be linked with this.

Thanks,
Alice

### CODE REVIEW

**Comment by alice-hannah** [comment]
Just a heads up that I've changed my mind and am trying to write the dbplyr stuff in SQL - will let you know when I'm done!

**Comment by alice-hannah** [comment]
Ok, ready for review now. 

I'm also thinking it'd probably be better to move the sql queries to their own script and source this at the start of the data preparation script. Again, we can rejig structure once the data prep stuff has been merged.

Thanks,
Alice

**Comment by alice-hannah** [comment]
Hi Tina,
These scripts are also now ready for review.

The section of code in 01_create-data.R that caps length of stay at 182.5 may change - some SMR01 stays overlap or are embedded within an SMR04 stay and I want to ensure we are dealing with these accurately and not double counting bed days. However, I don't think this should hold us back with review as I can come back and change thislater.

Thanks,
Alice

**Comment by Tina815** [comment]
Okay thanks Alice! I'm looking at the RMarkdown code so will come to this after finishing.

File: functions/sql_queries.R [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+       
+       # Financial year of death
+       "case when extract(month from date_of_death) between 1 and 3 ",
+       "then concat(to_char(add_months(date_of_death, -12), 'yyyy'), ",
+       "concat('/', to_char(date_of_death, 'yy'))) ",
+       "else concat(to_char(date_of_death, 'yyyy'), ",
+       "concat('/', to_char(add_months(date_of_death, 12), 'yy'))) ",
+       "end fy, ",
+       
+       # Quarter of death
```
Comment (Tina815) [inline functions/sql_queries.R]: Change it to "Financial quarter of death"?

File: code/00_setup-environment.R [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+postcode <- 
+  
+  read_rds(glue("{filepath}lookups/Unicode/Geography/",
+                "Scottish Postcode Directory/",
+                "Scottish_Postcode_Directory_2019_2.rds")) %>%
+  
+  clean_names() %>%
+  
+  select(pc7, ca2019, ca2019name, ca2018, hb2019, hb2019name,
```
Comment (Tina815) [inline code/00_setup-environment.R:87]: Do you need to include hb2018?

File: code/00_setup-environment.R [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+         hscp2019, hscp2019name, hscp2018, ur8_2016, 
+         data_zone2011)
+  
+            
+simd     <- 
+  
+  read_rds(glue("{filepath}lookups/Unicode/Deprivation/",
+                "postcode_2019_1.5_simd2016.rds")) %>%
```
Comment (Tina815) [inline code/00_setup-environment.R]: We will release postcode_2019_2_simd2016.rds this week so you can change to that.

File: code/00_setup-environment.R (hunk 2 above)
Comment (alice-hannah) [reply code/00_setup-environment.R:87]: The indicator is also included in a Source dashboard that breaks the data down by quarter and locality. I've been advised that some postcodes for 2019 variables have moved CA and HSCP, but not datazone, which is required to match on locality. Using the 2018 variables ensures these still match. Health board isn't required for this output, hence why this isn't included.

Is this something that will be fixed with the new postcode file you mention?

Comment (Tina815) [reply code/00_setup-environment.R:87]: As long as I know why you don't need hb2018 then that's fine. 

No I only said we would add geography names to the new postcode file and we've done that. 

Using 2018 geographies sound fine with me, as there's no population or geography mapping affected for 2018 boundary change.

File: code/01_create-data.R [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+  # For stays spanning this date, fix admission date to six months before death
+  mutate(admission_date = if_else(admission_date < six_months & discharge_date >= six_months,
+                                  six_months,
+                                  admission_date)) %>%
+  
+  # Select only stays within last six months of life
+  filter(admission_date >= six_months) %>%
+  
+  # Remove records where admission date is after date of death
+  filter(admission_date <= date_of_death) %>%
```
Comment (Tina815) [inline code/01_create-data.R]: Should be admission_date > date_of_death, since it's after date of death?

File: code/01_create-data.R [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+  # Calculate date six months before death
+  mutate(six_months = date_of_death - days(183)) %>%
+  
+  # For stays spanning this date, fix admission date to six months before death
+  mutate(admission_date = if_else(admission_date < six_months & discharge_date >= six_months,
+                                  six_months,
+                                  admission_date)) %>%
+  
+  # Select only stays within last six months of life
+  filter(admission_date >= six_months) %>%
```
Comment (Tina815) [inline code/01_create-data.R]: Based on the last step for fixing the admission date, can this line of code be taken out as admission_date will always be the same or after six_months?

File: code/01_create-data.R [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+  left_join(deaths, smr, by = "link_no") %>%
+  
+  group_by(fy, quarter, hb, ca, hscp, ca2018, hscp2018, locality,
+           simd, simd_15, sex, age_grp, urban_rural) %>%
+  
+  summarise(los = sum(los, na.rm = TRUE),
+            deaths = n())
+
+
+saveRDS(final, here("data", glue("{pub_date}_base-file.rds")))
```
Comment (Tina815) [inline code/01_create-data.R]: Shall we add a line of code to create a "data" folder in the working directory? Probably having it in 00_setup-environment.R?

**Review by Tina815 — COMMENTED** [review]
Hi Alice, I've reviewed the code and they look very good to me. Just left a couple of comments.

I noticed alice branch is 44 commits ahead, 29 commits behind master. I guess it's because I merged the rmarkdown branch to the master, but it's not in the alice branch. Will it create a merging conflict if we merge alice to master later?

File: code/01_create-data.R (hunk 5 above)
Comment (alice-hannah) [reply code/01_create-data.R]: We need to keep this line as the earlier step only fixes admission date where discharge date is within the last six months. There will also be stays that start and end more than six months before death, and this line gets rid of these.

File: code/01_create-data.R (hunk 4 above)
Comment (alice-hannah) [reply code/01_create-data.R]: This is what we want to remove so the filter condition defines what we want to keep, admission dates before date of death.

Comment (Tina815) [reply code/01_create-data.R]: That makes sense! Sorry I missed reading the word "remove"!

File: code/01_create-data.R (hunk 5 above)
Comment (Tina815) [reply code/01_create-data.R]: This makes sense after I drew a picture of the timeline:-)

**Comment by alice-hannah** [comment]
Thanks for reviewing Tina. I've made your suggested changes.

As I mentioned in my earlier comment, I'd also come across an issue where length of stay can be over counted due to overlapping SMR01 and SMR04 records. I've now fixed this and would be grateful if you could also review this section of code before merging?

Thanks, you're a star ⭐️ 
Alice

p.s. There shouldn't be any conflicts when merging as the markdown and code folders are separate. We've not made any changes to files in common between the branches

**Comment by Tina815** [comment]
Thanks Alice. You've made great progress!

I will try to review them this Wednesday. If not enough time I will continue next Tuesday.

File: functions/sql_queries.R [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+    # Select records in reporting period (and six months before)
+    "and s.discharge_date between ",
+    "to_date({shQuote(extract_start_smr, type = 'sh')}, 'yyyy-mm-dd') ",
+    "and to_date({shQuote(extract_end, type = 'sh')}, 'yyyy-mm-dd') ",
+    
+    # Exclude external causes of death
+    "and {{fn left(d.underlying_cause_of_death, 3)}} not in ",
+    "({paste0(shQuote(external_causes, type = 'sh'), collapse = ',')}) ",
+
+    "and ({{fn left(cause_of_death_code_0, 3)}} is null or ",
```
Comment (Tina815) [inline functions/sql_queries.R]: Just want to double check do we not need "d." in front of cause_of_death_code_0?

**Comment by Tina815** [comment]
Hi Alice, 

I've reviewed the latest code and they look good to me, especially the clever way to handle the overlap stays between SMR01 and SMR04! When I was trying to understand the method, I found it would be useful to know the data source (whether from SMR01 or SMR04) to help me understand the whole journey of the patient. Would you like to add a column say recid = "01B" for smr01 (01 and 50 combined), and "04B" for smr04? I did it myself when checking the code. 

I also left a question for the SQL query. 

I'm happy with the other things and ready to merge it into the master so that we've got the basefile!

**Comment by alice-hannah** [comment]
Thanks Tina - I've made your suggested changes. I'm happy to do the merge in case there are any conflicts (I think it'll just be with the gitignore file), but can you approve your review so I can do this please?

File: functions/sql_queries.R (hunk 7 above)
Comment (Tina815) [reply functions/sql_queries.R]: Thanks for adding them!

**Review by Tina815 — APPROVED** [review]

**Pull request merged**